### PR TITLE
Variable time step algorithm for Navier

### PR DESCRIPTION
**Pull request by jandrej**
This PR implements the variable step size IMEX method described in [1]. Usage is demonstrated in a new example run which is a modified version of its constant step size example.

[1] D. Wang, S.J. Ruuth (2008) Variable step-size implicit-explicit linear multistep methods for time-dependent partial differential equations

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1848](https://github.com/mfem/mfem/pull/1848) | @jandrej | @tzanio | @jamiebramwell + @kmittal2 + @cjvogl | 11/01/20 | 11/17/20 | 11/22/20 | |

### CODE REVIEW

File: miniapps/navier/navier_kovasznay_vs.cpp [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//
+// Navier Kovasznay example with variable time step
```
[the remark this answers is not in the document]
Comment (jandrej) [reply miniapps/navier/navier_kovasznay_vs.cpp:12]: Just like the specialized examples, where a diff can show "what changes are needed to use a specific feature". At least that was my intention.

File: miniapps/navier/navier_solver.hpp [hunk 2]
```diff
@@ -160,7 +160,31 @@ class NavierSolver
...
+    * The method and parameter choices are based on [1].
+    *
+    * [1] D. Wang, S.J. Ruuth (2008) Variable step-size implicit-explicit
+    * linear multistep methods for time-dependent partial differential
+    * equations
+    */
+   void Step(double &time, double dt, int cur_step, bool provisional = false);
+
+   /// Return a pointer to the provisional velocity ParGridFunction.
+   ParGridFunction *GetProvisionalVelocity() { return &un_next_gf; }
```
Comment (kmittal2) [inline miniapps/navier/navier_solver.hpp:187]: Mark this and other similar methods with `const`?

File: miniapps/navier/navier_kovasznay_vs.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmittal2) [reply miniapps/navier/navier_kovasznay_vs.cpp:12]: Okay, fair enough. What do you think @tzanio?

**Review by kmittal2 — APPROVED** [review]
Can you please update CHANGELOG?

Besides a couple minor comments, the PR looks good to me.

**Comment by tzanio** [comment]
@jamiebramwell and @cjvogl, can you please review?

**Review by jamiebramwell — APPROVED** [review]
This is a great addition!

File: miniapps/navier/navier_solver.cpp [hunk 3]
```diff
@@ -288,57 +292,80 @@ void NavierSolver::Setup(double dt)
...
-         Vector diag_pa(vfes->GetTrueVSize());
-         H_form->AssembleDiagonal(diag_pa);
-         HInvPC = new OperatorJacobiSmoother(diag_pa, vel_ess_tdof);
-         HInv->SetPreconditioner(*HInvPC);
-      }
-      else
-      {
-         HInv->SetOperator(*H);
-      }
+      HInv->SetOperator(*H);
```
Comment (cjvogl) [inline miniapps/navier/navier_solver.cpp:353]: Maybe nitpicking here, especially given that this structure is inherited from master, but could move this line before `if` and then remove `else`

Comment (jandrej) [reply miniapps/navier/navier_solver.cpp:353]: No you are right, the check is unnecessary now. Maybe I should check if the `dt` history is the same, but that could complicate things.

**Review by cjvogl — APPROVED** [review]
I'm not all too familiar (read: not at all familiar) with the Navier Stokes miniapp, but the code looks sound, runs, and produces an eye-pleasing picture!

**Comment by cjvogl** [comment]
> @jamiebramwell and @cjvogl, can you please review?

Sorry, I mistakenly put the Merge deadline as the Approval deadline in my calendar

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by jandrej** [comment]
@tzanio I updated the branch with changes that were requested from the reviewers after you merged in next.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

File: miniapps/navier/navier_kovasznay_vs.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tzanio) [reply miniapps/navier/navier_kovasznay_vs.cpp:12]: I think that's OK, these codes are meant more as examples and code duplication is not bad for those.

**Review by tzanio — APPROVED** [review]
Looks good, thanks for keep pushing on this @jandrej !

@jamiebramwell, @kmittal2 and @cjvogl -- thank you for reviewing!

**Pull request merged**